### ISSUE TITLE
Add response interface and implementations

### ISSUE THREAD

**Issue by wsanchez** [wsanchez/txrequest#1]
[`IRequest`](http://twistedmatrix.com/documents/current/api/twisted.web.iweb.IRequest.html) encapsulates both a request and the response to that request.

If we are to replace `IRequest`, therefore, we need a response API.

**Comment by wsanchez** [comment]
I don't think that `IHTTPRequest` should follow `IRequest`'s lead here by trying to put both concepts into one object, so I think we want an `IHTTPResponse` and an `HTTPResponseFromIRequest` wrapper for conversion.

**Comment by glyph** [comment]
It's worth looking at as-yet purely client-side IResponse here, I think.

**Comment by wsanchez** [comment]
Ah, yes, thanks; I forgot this existed.

Though… I'm not seeing much here that seems useful…

 * `version`: The protocol version is a function of the transport, not the message, and the classes we're defining here are messages not coupled to a transport.  Meaning I should be able to create an `HTTPResponse` without specifying the protocol version, and that response should be deliverable in both HTTP 1.1 and HTTP 2.

 * `code`: Yeah, need that.

 * `phrase`: I don't see why that needs to be a feature of the response class. That said, I don't understand why HTTP has reason phrases at all; it seems like a stupid attempt at making them localizable.

 * `headers`: Yup.

 * `length`: I can't tell from the docs if this is bound to the `Content-Length` header or a separate way to specify the length.  If the latter, this seems like a bad idea.  I have to put some thought into where the "friendly" headers API should go (I feel like that should be separate from the core API, and extensible without having to subclass).

 * `request`: No sure that belongs here any more than the response belongs in the request.  My instinct is that if you want to associate a request and a response, you should do that elsewhere.

 * `previousResponse`: Same instinctual reaction as with `request`.

 * `deliveryBody`: This would be replaced by `bodyAsFount`, presumably…

**Comment by wsanchez** [comment]
So… my thinking for `IHTTPResponse` is that it's the same as `IHTTPRequest`, but removing the `method` and `uri` attributes and adding a `status` attribute.

Common between them is the `headers` attribute, and the `bodyAsFount` and `bodyAsBytes` methods.  This can be moved into a new parent interface, `IHTTPMessage`, using terminology from [RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html), in which an HTTP message is either a request or response, and a generic HTTP message consists of headers and a body.

**Comment by wsanchez** [comment]
#2 does the above.

Fitting this into Klein is a bit of getting a square peg into a round hole.  `IHTTPResponse`, wants to fully encapsulate a response, but Klein splits the response headers into a thing you twiddle on the request object and *separately*, the body is a thing you return:

```python
    @route(...)
    def foo(self, _request: IRequest) -> KleinRenderable:
        request = HTTPRequestFromIRequest(_request)
        response = HTTPResponseFromIRequest(_request)

        ...logic and stuff...

        # Set status and some headers
        response.setStatus(200)
        response.headers.addValue(...)

        response.setBody(...some bytes or a Fount...)

        # Convert a response to bytes or a deferred that drains into request.write
        return responseAsKleinRenderable(response)
```

Perhaps a better option is something that looks like this:

```python
    @route(...)
    def foo(self, _request: IRequest) -> KleinRenderable:
        request = HTTPRequestFromIRequest(_request)

        ...logic and stuff...

        # Build some headers
        headers = MutableHTTPHeaders()
        headers.addValue(...)

        response = HTTPResponse(
            status=200,
            headers=headers,
            body=...some bytes or a Fount...,
        )

        # Utility that shovels a response object headers into an `IRequest` and returns the body.
        return responseIntoIRequest(response, _request)
```

I like the latter because it puts all of the conversion "at the edges".  And evolving Klein into handling this natively seems doable; we can probably build a thing that wraps an `IRequest` such that it is *both* an `IRequest` and an `IHTTPRequest`, and Klein can be adapted to handle an `IHTTPResponse` as a return value.

**Comment by wsanchez** [comment]
Merged #2.

**Comment by wsanchez** [comment]
Opened #3 for thinking about Klein
